Monika After Story/game/script-affection.rpy

```diff
@@ -2078,7 +2080,7 @@ label monika_affection_nickname:
                             m 2ektsc "...You didn't have to be so mean."
                             m 2dftdc "That really hurt, [player]."
 
-                            if  mas_getEV('mas_apology_bad_nickname').shown_count == 2:
+                            if mas_checkEV("mas_apology_bad_nickname", lambda x: EV_SHOWN_COUNT_EQUAL_TO(x, 2)):
```

Comment:
Cases like these make me question exactly how useful it is to do this. For example, lets compare the previous code 
 (with none checking added):
```python
$ bad_nick_ev = mas_getEV("mas_apology_bad_nickname")
if bad_nick_ev is not None and bad_nick_ev.shown_count == 2:
   # ...
```
In the above code (lets call it the `getEV` case), at most the following happens:
1. a db lookup is performed
2. None conditional check
3. getattr retrieval `shown_count`
4. `== 2` conditional check

And the resulting bytecode:
```
 15           0 LOAD_GLOBAL              0 (mas_getEV)
              3 LOAD_CONST               1 (u'mas_apology_bad_nickname')
              6 CALL_FUNCTION            1
              9 STORE_FAST               0 (bad_nick_ev)

 16          12 LOAD_FAST                0 (bad_nick_ev)
             15 LOAD_CONST               0 (None)
             18 COMPARE_OP               9 (is not)
             21 POP_JUMP_IF_FALSE       48
             24 LOAD_FAST                0 (bad_nick_ev)
             27 LOAD_ATTR                2 (shown_count)
             30 LOAD_CONST               2 (2)
             33 COMPARE_OP               2 (==)
             36 POP_JUMP_IF_FALSE       48
```

For the `checkEV` case, the following happens:
1. create `lambda` function of a lambda function, which also means compile python bytecode for this function
2. db lookup is performed
3. None check conditional 
4. execute lambda function - includes calling two functions (the outer lambda and the constant lambda)

And the resulting bytecode for this: (I had to combine bytecodes with some of the individual functions since `dis` lib wont show it all at once)
```
 20           0 LOAD_GLOBAL              0 (mas_checkEV)
              3 LOAD_CONST               1 (u'mas_apology_bad_nickname')
              6 LOAD_CONST               2 (<code object <lambda> at 061B0380, file "game/outputcapture.rpy", line 20>)
              9 MAKE_FUNCTION            0
             12 CALL_FUNCTION            2

                    0 LOAD_GLOBAL              0 (mas_getEV)
                    3 LOAD_FAST                0 (ev_label)
                    6 CALL_FUNCTION            1
                    9 STORE_FAST               2 (ev)

                   12 LOAD_FAST                2 (ev)
                   15 LOAD_CONST               1 (None)
                   18 COMPARE_OP               8 (is)
                   21 POP_JUMP_IF_FALSE       28

                   24 LOAD_GLOBAL              2 (False)
                   27 RETURN_VALUE        

                   28 LOAD_FAST                1 (predicate)
                   31 LOAD_FAST                2 (ev)
                   34 CALL_FUNCTION            1

                        0 LOAD_GLOBAL              0 (EV_SHOWN_COUNT_EQUAL_TO)
                        3 LOAD_FAST                0 (x)
                        6 LOAD_CONST               1 (2)
                        9 CALL_FUNCTION            2

                            0 LOAD_FAST                0 (ev)
                            3 LOAD_ATTR                0 (shown_count)
                            6 LOAD_FAST                1 (sc)
                            9 COMPARE_OP               2 (==)
                           12 RETURN_VALUE        

                       12 RETURN_VALUE        

                   37 RETURN_VALUE        

             15 POP_JUMP_IF_FALSE       27
```

Comparing the bytecode, we can see that there are significantly more operations when using `mas_checkEV` in this case. Here I'm comparing the max operations we could encounter for an expected, non-None case. This also doesn't take into the account the fact that the lambda function has to be created every time we are executing this check. 

| op | getEV and None check | checkEV |
| --- | --- | --- |
| loads | 7 | 15 |
| calls | 1 | 4 |

Now, it's reasonable to use less efficient code in cases where performance isn't a problem and the less efficient code is more readable. However, I'm not so sure the `checkEV` usage here is more readable. In the `getEV` case, it's easy to see that we're grabbing some object, checking if it exists, then checking a property on that object. In the `checkEV` case, you'd have to understand that the lambda function will be executed on the object if it exists. Yes, the lambda constant used here is easy to understand, but it's also verbose - it means the same thing as `== 2` yet is like 10 times as long. At most, both `getEV` and `checkEV` have equal amount of readability - which puts the `checkEV` code at a disadvantage because it is definitely less efficient.

Additionally for other devs, using lambda here seems unnatural if you know that the ev functions work with Event objects. If you want to check a property on an object, it's normal to get the object then check the property - not build a function that does the check and pass it to another function.

Earlier, when I looked over the `mas_checkEV` function, I assumed the predicate option is for specific custom cases - rare cases. I don't think we should be defaulting to the `checkEV` function as the main way to do certain checks. If anything, the constants that you've written out in `event-handler` should really just be their own functions. The property-specific ones should have getter wrappers and the actual condition shouldn't be abstracted for clarity, unless its an obvious check. 

I propose these functions be created:
* `mas_isPoolEV`
* `mas_isRandomEV`
* `mas_isUnlockedEV`
* `mas_isSeenEV`
* `mas_getEV_last_seen`
* `mas_getEV_shown_count`
* `mas_inRulesEV`

All these functions could use `mas_getEVPropValue`. Defaulting should be a param for the `getEVs` that return properties.

We can keep `checkEV` around for the custom cases, but when we find ourselves constantly doing the same property check over and over, then we should make a specific function for it.

Reply:
I think I have a solution to the assign/modify issue. Should make it more efficient and more general to use overall.